pkg/log/locallogger/locallogger.go

```diff
@@ -69,5 +194,122 @@ func filterResults(keyvals ...interface{}) {
 			}
 		}
 	}
+}
+
+// hashKeyvals creates a hash of the key-value pairs for deduplication
+// We exclude timestamp and caller fields since they change for identical log content
+func (ll *localLogger) hashKeyvals(keyvals ...interface{}) string {
+	return hashKeyValuePairs(keyvals...)
+}
+
+// hashKeyValuePairs creates a consistent hash from key-value pairs, excluding timestamp fields
+func hashKeyValuePairs(keyvals ...interface{}) string {
+	hasher := sha256.New()
+
+	// Collect non-excluded key-value pairs
+	var pairs []string
+	for i := 0; i < len(keyvals); i += 2 {
+		if i+1 >= len(keyvals) {
+			break
+		}
+
+		key := fmt.Sprintf("%v", keyvals[i])
+
+		// Skip excluded fields for content-based deduplication
+		if excludedHashFields[key] {
+			continue
+		}
+
+		value := fmt.Sprintf("%v", keyvals[i+1])
+		pairs = append(pairs, key+":"+value)
+	}
+
+	// Sort pairs for consistent hashing
+	sort.Strings(pairs)
+
+	// Hash the sorted pairs
+	for _, pair := range pairs {
+		hasher.Write([]byte(pair + ";"))
+	}
+
+	return fmt.Sprintf("%x", hasher.Sum(nil))
+}
+
+// shouldSkipDuplicate checks if this log message is a duplicate and should be skipped
+// Returns (skip, duplicateCount) where skip indicates if the message should be skipped
+// and duplicateCount is the number of times this message has been seen
+func (ll *localLogger) shouldSkipDuplicate(hash string) (bool, int) {
+	ll.dedupMutex.Lock()
+	defer ll.dedupMutex.Unlock()
+
+	now := time.Now()
+
+	// Perform cleanup if needed
+	if now.Sub(ll.lastCleanup) > cleanupInterval {
+		ll.cleanupCacheUnsafe(now)
+		ll.lastCleanup = now
```

Comment:
I think we should set `ll.lastCleanup = now` inside `ll.cleanupCacheUnsafe` -- then we don't have to worry about data races with the `ll.lastCleanup` field if we ever have more routines calling `ll.cleanupCacheUnsafe`, because it will be effectively protected by `ll.dedupMutex`. Plus, responsibility-wise, I think it makes sense for the cache-cleaning routine to own the cache clean timestamp.